pkg/services/meta/containers.go

```diff
@@ -587,3 +593,121 @@ func (s *containerStorage) handleNewEpoch(e uint64) error {
 
 	return errors.Join(mptErr, dbErr)
 }
+
+func expandChildren(st storage.Store, rootOIDs []byte) []byte {
+	if len(rootOIDs) == 0 {
+		return nil
+	}
+
+	// sorting before every SEEK is used for exact single operation for every
+	// children searching subroutine; search is done in 3 steps:
+	//  1. root -> last/link object search
+	//  2. last/link -> first part ID search
+	//  3. first part ID -> all children search
+
+	var childrenWithParentID [][]byte
+	var rng storage.SeekRange
+	rng.Prefix = slices.Concat([]byte{attrPlainToOIDIndex}, []byte(objectsdk.FilterParentID), object.AttributeDelimiter)
+	rng.Start = make([]byte, oid.Size)
+
+	rootsSorted := slices.SortedFunc(slices.Chunk(rootOIDs, oid.Size), bytes.Compare)
+	lastRoot := rootsSorted[len(rootsSorted)-1]
+	copy(rng.Start, rootsSorted[0])
+
+	st.Seek(rng, func(k, _ []byte) bool {
+		currRoot := k[len(rng.Prefix) : len(rng.Prefix)+oid.Size]
+		if len(rootsSorted) == 0 || bytes.Compare(currRoot, lastRoot) > 0 {
+			return false
+		}
+		for i := 0; i < len(rootsSorted); i++ {
+			switch bytes.Compare(currRoot, rootsSorted[i]) {
+			case -1:
+				return true
+			case +1:
+				rootsSorted = rootsSorted[i+1:]
+				i = -1
+				continue
+			case 0:
+				childrenWithParentID = append(childrenWithParentID, slices.Clone(k[len(rng.Prefix)+oid.Size+1:]))
+				rootsSorted = rootsSorted[i+1:]
+				return true
+			}
+		}
+
+		return true
+	})
+
+	if len(childrenWithParentID) == 0 {
+		// all objects are roots, no additional children
+		return rootOIDs
+	}
+
+	firstPartOIDs := make([][]byte, 0, len(childrenWithParentID))
+	rng.Prefix = []byte{oidToAttrIndex}
+	copy(rng.Start, childrenWithParentID[0])
+
+	slices.SortFunc(childrenWithParentID, bytes.Compare)
+	st.Seek(rng, func(k, _ []byte) bool {
+		if len(childrenWithParentID) == 0 {
+			return false
+		}
+		if !bytes.HasPrefix(k[1+oid.Size:], []byte(objectsdk.FilterFirstSplitObject)) {
+			return true
+		}
+		currChild := k[1 : 1+oid.Size]
+		for i := 0; i < len(childrenWithParentID); i++ {
+			switch bytes.Compare(currChild, childrenWithParentID[i]) {
+			case -1:
+				return true
+			case +1:
+				// should never happen, it means we have info
+				// about child object with parent ID but child
+				// object does not have first part ID, nothing
+				// can be done, just skip
```

Comment:
log?

Reply:
what this log should say? like we have some garbage or unknown db state?

Reply:
the same as the comment. It's worth leaving at least one way to detect this case. Log costs nothing for this

Reply:
> Log costs nothing for this

it is a log record inside an opened db tx that is asked to be idempotent (ok, it is a `Storage` interface but we all know what it means in practice)

fixed